### PR TITLE
Disabled controls should lose focus and don't accept keyboard events

### LINKED ISSUE #7294: Can't properly disable focused controls
**Describe the bug**
Some key events are still fully functional in disabled+focused controls. Probably related to https://github.com/AvaloniaUI/Avalonia/issues/6957, but [the PR](https://github.com/AvaloniaUI/Avalonia/pull/7206) does not solves other controls. With Fluent theme, control style don't get updated. But I can see in the debugger that `IsEnabled` is already set to false.

**To Reproduce**
1. Make `TextBox` focus
2. Disable the `TextBox`
3. Start typing..

[FocusedControlsIssue.zip](https://github.com/AvaloniaUI/Avalonia/files/7797011/FocusedControlsIssue.zip)

**Expected behavior**
Disabled controls should not accept key events.

**Screenshots**

https://user-images.githubusercontent.com/57059775/147838556-cb1a0b3b-227a-4784-ae87-041aa0ad19d0.mp4

**Desktop (please complete the following information):**

- OS: Debian (dwm)
- Version: 0.10.10

### PR DESCRIPTION
**Pull request by Takoooooo**
## What does the pull request do?
Previously, if you would disable a focused control from the code it will still stay focused and it will be able to receive events.
In case of TextBox it will lead to TextBox staying Enabled while you won't move the focus. But the same thing may happen to other controls as well. Big thanks to @luthfiampas because i used his solution here and it looks good to me.

## What is the current behavior?
Currently,when user sets control IsEnabled property to false this control will loase focus and focus wouldn't be aquired until it will stay disabled. Thats how it works in WPF - "In WPF a control is no longer interactive when it is disabled (IsEnabled=false). The focus will be lost. This can also lead to a null value for the current focused element when the current keyboard focus "holder" gets disabled by a binding etc."

## Fixed issues
Fixes #7294

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027077-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

File: src/Avalonia.Base/Input/InputElement.cs [hunk 1]
```diff
@@ -442,6 +443,23 @@ private set
...
+                        FocusManager.Instance?.Focus(null);
+                    }
+                    else
+                    {
+                        _restoreFocus = false;
+                    }
+                }
+                else if (IsEffectivelyEnabled && _restoreFocus)
+                {
+                    FocusManager.Instance?.Focus(this);
```
Comment (danwalmsley) [inline src/Avalonia.Base/Input/InputElement.cs:461]: what should happen in this scenario:

1. control is disabled, looses focus.
2. another control gets focused
3. the original control is enabled again...

At step 3 will the original control steal the focus?

**Review by danwalmsley — COMMENTED** [review]
Question

**Review by grokys — CHANGES_REQUESTED** [review]

File: src/Avalonia.Base/Input/InputElement.cs (hunk 1 above)
Comment (Takoooooo) [reply src/Avalonia.Base/Input/InputElement.cs]: Should be fixed now.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027199-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

**Comment by Takoooooo** [comment]
@grokys robloo suggests to add AllowDisabledFocus property because it may be needed for accessibility,what do you think? I assume i may do that

**Comment by grokys** [comment]
> robloo suggests to add AllowDisabledFocus property because it may be needed for accessibility,what do you think? I assume i may do that

If we want to add that, probably should be in a separate PR. I think its use is pretty niche though.

**Review by grokys — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027262-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

**Comment by klorman** [comment]
Is there any update on the AllowDisabledFocus feature discussed here? Are there other ways to keep focus on disabled controls?

**Comment by cla-avalonia** [comment]
@Takoooooo,

Please read the following Contributor License Agreement (CLA). If you agree with the CLA, please reply with the following:
```
@cla-avalonia agree
```

<details>
    <summary>Contributor License Agreement</summary>
    
### Contribution License Agreement

This Contribution License Agreement ( **“Agreement”** ) is agreed to by the party signing below ( **“You”** ),
and conveys certain license rights to the AvaloniaUI OÜ ( **“AvaloniaUI OÜ”** ) for Your contributions to
AvaloniaUI OÜ open source projects. This Agreement is effective as of the latest signature date below.

**1. Definitions.**

**“Code”** means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to AvaloniaUI OÜ under this Agreement.

**“Project”** means any of the projects owned or managed by AvaloniaUI OÜ and offered under a license
approved by the Open Source Initiative (www.opensource.org).

**“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”

**“Submission”** means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.

**2. Your Submission.** You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.

**3. Originality of Work.**  You represent that each of Your Submissions is entirely Your 
original work. Should You wish to Submit materials that are not Your original work, 
You may Submit them separately to the Project if You (a) retain all copyright and 
license information that was in the materials as you received them, (b) in the 
description accompanying your Submission, include the phrase &quot;Submission 
containing materials of a third party:&quot; followed by the names of the third party and any 
licenses or other restrictions of which You are aware, and (c) follow any other 
instructions in the Project&#39;s written guidelines concerning Submissions.

**4. Your Employer.** References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.

**5. Licenses.**

**a. Copyright License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly
or indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable
license in the Submission to reproduce, prepare derivative works of, publicly display, publicly perform,
and distribute the Submission and such derivative works, and to sublicense any or all of the foregoing
rights to third parties.

**b. Patent License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly or
indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license
under Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.

**c. Other Rights Reserved.** Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

**6. Representations and Warranties.** You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3 ). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6 , THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.

**7. Notice to AvaloniaUI OÜ.** You agree to notify AvaloniaUI OÜ in writing of any facts or
circumstances of which You later become aware that would make Your representations in this
Agreement inaccurate in any respect.

**8. Information about Submissions.** You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.

**9. Governing Law/Jurisdiction.** This Agreement is governed by the laws of the Republic of Estonia, and
the parties consent to exclusive jurisdiction and venue in the courts sitting in Talinn,
Estonia. The parties waive all defenses of lack of personal jurisdiction and forum non-conveniens.

**10. Entire Agreement/Assignment.** This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by AvaloniaUI OÜ.


*AvaloniaUI OÜ dedicates this Contribution License Agreement to the public domain according to the Creative Commons CC0 1.*

</details>

**1** out of **2** committers have signed the CLA.
- [x] @maxkatz6
- [ ] @Takoooooo

**Pull request merged**